lib/Sema/CSStep.h

```diff
@@ -607,21 +630,34 @@ class TypeVariableStep final : public BindingStep<TypeVarBindingProducer> {
         << InitialBindings.size() << " initial bindings\n";
   }
 
+private:
+  /// Checks whether any of the attempted bindings so far have produced a
+  /// solution without needing to add fixes or bind to an unavailable decl.
+  bool solvedAnyWithoutFixes() const {
+    if (!AnySolved)
+      return false;
+
+    // Check if any of the bindings introduced fixes or bound unavailable
+    // overloads.
+    auto scoreDelta = *BestScore - OriginalScore;
+    return scoreDelta.Data[SK_Fix] == 0 && scoreDelta.Data[SK_Unavailable] == 0;
+  }
+
 protected:
   bool attempt(const TypeVariableBinding &choice) override;
 
   bool shouldSkip(const TypeVariableBinding &choice) const override {
     // If this is a defaultable binding and we have found solutions,
     // don't explore the default binding.
-    return AnySolved && choice.isDefaultable();
+    return solvedAnyWithoutFixes() && choice.isDefaultable();
```

Comment:
It might be more interesting to record collection element types to be holes by default instead of allowing to try `Any` because in diagnostics we'd like to clearly indicate that something was not inferred via `_`.

Reply:
Hmm, I think we still need the defaultable `Any` constraint to allow the following:

```swift
func foo<T>(_ x: T) {}
foo([])
```

Actually, on second thought, I think I'll revert this change to `shouldSkip` for now. I initially added it to be consistent with the other two methods, but I can't really think of any useful diagnostics we'd actually get from it.

Reply:
I think for such cases we should be defaulting generic parameter to a hole and propagating it so I'd suggest we drop changes to `shouldSkip`.